plugins/modules/vmware_guest_vgpu.py

```diff
@@ -0,0 +1,394 @@
+        """
+        Remove vGPU profile of virtual machine
+        Args:
+            vm_obj: Managed object of virtual machine
+            vgpu_VirtualDevice_obj: vGPU profile object holding its facts
+            vgpu_prfl: vGPU profile name
+        Returns: Operation results
+        """
+        changed = False
+        failed = False
+        vm_current_vgpu_profile = self._get_vgpu_profile_in_the_vm(vm_obj)
+        if vgpu_prfl in vm_current_vgpu_profile:
+            vdspec = vim.vm.device.VirtualDeviceSpec()
+            vmConfigSpec = vim.vm.ConfigSpec()
+            vdspec.operation = "remove"
+            vdspec.device = vgpu_VirtualDevice_obj
+            vmConfigSpec.deviceChange.append(vdspec)
+
+            try:
+                task = vm_obj.ReconfigVM_Task(spec=vmConfigSpec)
+                wait_for_task(task)
+                changed = True
+                return changed, failed
+            except Exception as exc:
+                failed = True
+                self.module.fail_json(
+                    msg="Failed to delete vGPU profile"
+                    " '%s' from vm %s." % (vgpu_prfl, vm_obj.name),
+                    detail=exc.msg,
+                )
+        return changed, failed
+
+    def _vgpu_present(self, vm_obj):
+        """
+        Add vGPU profile to virtual machine.
+        Args:
+            vm_obj: Managed object of virtual machine
+        Returns: Operation results and vGPU facts
+        """
+        result = {}
+        vgpu_prfl = self.params["vgpu"]
+        vgpu_profile_name = self._get_vgpu_profiles_name(vm_obj, vgpu_prfl)
+        if vgpu_profile_name is None:
+            self.module.fail_json(
+                msg="vGPU Profile '%s'" " does not exist." % vgpu_prfl
+            )
+
+        changed, failed = self._add_vgpu_profile_to_vm(
+            vm_obj, vgpu_profile_name, vgpu_prfl
+        )
+        vgpu_fact = self._gather_vgpu_profile_facts(vm_obj)
+        result = {"changed": changed, "failed": failed, "vgpu": vgpu_fact}
+        return result
+
+    def _add_vgpu_profile_to_vm(self, vm_obj, vgpu_profile_name, vgpu_prfl):
+        """
+        Add vGPU profile of virtual machine
+        Args:
+            vm_obj: Managed object of virtual machine
+            vgpu_profile_name: vGPU profile object name from ESXi server list
+            vgpu_prfl: vGPU profile name
+        Returns: Operation results
+        """
+        changed = False
+        failed = False
+        vm_current_vgpu_profile = self._get_vgpu_profile_in_the_vm(vm_obj)
+        if self.params["force"] or vgpu_prfl not in vm_current_vgpu_profile:
+            vgpu_p = vgpu_profile_name.vgpu
+            backing = vim.VirtualPCIPassthroughVmiopBackingInfo(vgpu=vgpu_p)
+            summary = "NVIDIA GRID vGPU " + vgpu_prfl
+            deviceInfo = vim.Description(summary=summary, label="PCI device 0")
+            hba_object = vim.VirtualPCIPassthrough(
```

Comment:
I haven't worked yet with vGPUs and don't have the opportunity to test this, so I have to look at the code and figure out if it makes sense to me. Generally, your code does... but I'm not sure if I understand why you used the name `hba_object` here. When I hear HBA, I think (SCSI or FC) host bus adapter but not GPU.

Maybe HBA has a meaning when working with vGPUs that I don't know. If not, can you please use a more appropriate term here?

Apart from this, your code looks good to me.

Reply:
Hi @mariolenz, I have no idea why VMware is putting vGPU profiles under HBA but it is listed as an HBA object that is sitting in the same list as the SCSI and other adapters. So this is not your typical GPU, it is not done by passthrough, this is a VMware and Nvidia weird vGPU profile system, which I assume they decided the best place to list it was under HBA.

Thanks for reviewing my code. Honestly, I have not done much object coding for years now, I was mainly adapting code from other modules you guys wrote to get this working.

Reply:
@Medalibi Thanks for the explanation :-)